File: app/src/main/java/com/hover/stax/actions/ActionDropdownAdapter.java
package com.hover.stax.actions;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;

import com.hover.sdk.actions.HoverAction;
import com.hover.stax.R;
import com.hover.stax.databinding.StaxSpinnerItemWithLogoBinding;
import com.hover.stax.utils.UIHelper;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.util.List;

import timber.log.Timber;

import static com.hover.stax.utils.Constants.size55;


public class ActionDropdownAdapter extends ArrayAdapter<HoverAction> {
    private final List<HoverAction> actions;

    public ActionDropdownAdapter(@NonNull List<HoverAction> actions, @NonNull Context context) {
        super(context, 0, actions);
        this.actions = actions;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View view, @NonNull ViewGroup parent) {
        HoverAction a = actions.get(position);
        ViewHolder holder;

        if (view == null) {
            StaxSpinnerItemWithLogoBinding binding = StaxSpinnerItemWithLogoBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            view = binding.getRoot();

            holder = new ViewHolder(binding);
            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }

        holder.setAction(a, getContext().getString(R.string.root_url));

        return view;
    }


    private static class ViewHolder implements Target {
        TextView id;
        ImageView logo;
        AppCompatTextView channelText;
        View divider;

        StaxSpinnerItemWithLogoBinding binding;

        private ViewHolder(StaxSpinnerItemWithLogoBinding binding) {
            this.binding = binding;

            logo = binding.serviceItemImageId;
            channelText = binding.serviceItemNameId;
            id = binding.serviceItemId;
            divider = binding.serviceItemDivider;
            divider.setVisibility(View.GONE);
        }

        @SuppressLint("SetTextI18n")
        private void setAction(HoverAction action, String baseUrl) {
            id.setText(Integer.toString(action.id));
            channelText.setText(action.toString());
            UIHelper.loadPicasso(baseUrl + action.to_institution_logo, size55, this);
        }

        @Override
        public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
            RoundedBitmapDrawable d = RoundedBitmapDrawableFactory.create(id.getContext().getResources(), bitmap);
            d.setCircular(true);
            logo.setImageDrawable(d);
        }

        @Override
        public void onBitmapFailed(Exception e, Drawable errorDrawable) {
            Timber.e(e);
        }

        @Override
        public void onPrepareLoad(Drawable placeHolderDrawable) {
        }
    }

    @Override
    public int getCount() {
        return actions.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }
}

File: app/src/main/java/com/hover/stax/actions/ActionDropdownAdapter.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.actions

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.widget.AppCompatTextView
import com.hover.sdk.actions.HoverAction
import com.hover.stax.R
import com.hover.stax.databinding.StaxSpinnerItemWithLogoBinding
import com.hover.stax.utils.UIHelper.loadImage

class ActionDropdownAdapter(val actions: List<HoverAction>, context: Context) : ArrayAdapter<HoverAction>(context, 0, actions) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val action = actions[position]
        val holder: ActionViewHolder

        if (view == null) {
            val binding = StaxSpinnerItemWithLogoBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            view = binding.root
            holder = ActionViewHolder(binding)
            view.tag = holder
        } else
            holder = view.tag as ActionViewHolder

        holder.setAction(action, context.getString(R.string.root_url))

        return view
    }

    override fun getCount(): Int = actions.size

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getItem(position: Int): HoverAction? = if (actions.isEmpty()) null else actions[position]

    class ActionViewHolder(val binding: StaxSpinnerItemWithLogoBinding) {

        private var id: TextView = binding.serviceItemId
        private var logo: ImageView = binding.serviceItemImageId
        private var channelText: AppCompatTextView = binding.serviceItemNameId

        fun setAction(action: HoverAction, baseUrl: String) {
            id.text = action.id.toString()
            channelText.text = action.toString()
            val logoUrl = baseUrl.plus(action.to_institution_logo)
            logo.loadImage(binding.root.context, logoUrl)
        }
    }
}